File: 2-strings/java/7-reverse-integer.java
public class Solution{
    public static int reverse(int x) {
        int result = 0;

        while(x != 0) {
            int reminder = x % 10;
            int temp = (result * 10) + reminder;

            if((temp-reminder) / 10 != result){
                return 0;
            }

            x /= 10;
            result = temp;
        }
        return result;
    }

    public static void main(String args[]){
        int x = 123;
        System.out.println(reverse(x));
    }
}

// Time Complexity: O(log(n))
// Space Complexity: O(1)

File: 2-strings/cpp/7-reverse-integer.cpp
#include <iostream>
using namespace std;

int reverse(int x) {
    long long result = 0;

    while (x != 0) {
        int reminder = x%10;
        int temp = (result*10) + reminder;
        if ((temp - reminder) / 10 != result) {
            return 0;
        }

        x /= 10;
        result = temp;
    }
    return result;
}
int main() {
    int x = 1534236469;
    cout << reverse(x) << endl;
    
    return 0;
}

// Time Complexity: O(log(n))
// Space Complexity: O(1)
